services/cargohold/src/CargoHold/AWS.hs

```diff
@@ -0,0 +1,187 @@
+useDownloadEndpoint :: Env -> Env
+useDownloadEndpoint e =
+  e & amazonkaEnv %~ AWS.override (setAWSEndpoint (e ^. amazonkaDownloadEndpoint))
+
+setAWSEndpoint :: AWSEndpoint -> AWS.Service -> AWS.Service
+setAWSEndpoint e = AWS.setEndpoint (_awsSecure e) (_awsHost e) (_awsPort e)
+
+newtype Amazon a = Amazon
+  { unAmazon :: ReaderT Env (ResourceT IO) a
+  }
+  deriving
+    ( Functor,
+      Applicative,
+      Monad,
+      MonadIO,
+      MonadThrow,
+      MonadCatch,
+      MonadMask,
+      MonadReader Env,
+      MonadResource
+    )
+
+instance MonadLogger Amazon where
+  log l m = view logger >>= \g -> Logger.log g l m
+
+instance MonadUnliftIO Amazon where
+  askUnliftIO = Amazon $ ReaderT $ \r ->
+    withUnliftIO $ \u ->
+      return (UnliftIO (unliftIO u . flip runReaderT r . unAmazon))
+
+instance AWS.MonadAWS Amazon where
+  liftAWS a = view amazonkaEnv >>= flip AWS.runAWS a
+
+mkEnv ::
+  Logger ->
+  -- | S3 endpoint
+  AWSEndpoint ->
+  -- | Endpoint for downloading assets (for the external world)
+  AWSEndpoint ->
+  -- | Bucket
+  Text ->
+  Maybe CloudFrontOpts ->
+  Manager ->
+  IO Env
+mkEnv lgr s3End s3Download bucket cfOpts mgr = do
+  let g = Logger.clone (Just "aws.cargohold") lgr
+  e <- mkAwsEnv g (setAWSEndpoint s3End S3.s3)
+  cf <- mkCfEnv cfOpts
+  return (Env g bucket e s3Download cf)
+  where
+    mkCfEnv (Just o) = Just <$> initCloudFront (o ^. cfPrivateKey) (o ^. cfKeyPairId) 300 (o ^. cfDomain)
+    mkCfEnv Nothing = return Nothing
+    mkAwsEnv g s3 =
+      AWS.newEnvWith AWS.Discover Nothing mgr
+        <&> set AWS.envLogger (awsLogger g)
+        <&> AWS.configure s3
+    awsLogger g l = Logger.log g (mapLevel l) . Logger.msg . toLazyByteString
+    mapLevel AWS.Info = Logger.Info
+    mapLevel AWS.Debug = Logger.Trace
+    mapLevel AWS.Trace = Logger.Trace
+    mapLevel AWS.Error = Logger.Debug
+
+execute :: MonadIO m => Env -> Amazon a -> m a
+execute e m = liftIO $ runResourceT (runReaderT (unAmazon m) e)
+
+data Error where
+  GeneralError :: (Show e, AWS.AsError e) => e -> Error
+
+deriving instance Show Error
+
+deriving instance Typeable Error
+
+instance Exception Error
+
+--------------------------------------------------------------------------------
+-- Utilities
+
+sendCatch :: AWSRequest r => r -> Amazon (Either AWS.Error (Rs r))
+sendCatch = AWST.trying AWS._Error . AWS.send
+
+send :: AWSRequest r => r -> Amazon (Rs r)
+send r = throwA =<< sendCatch r
+
+throwA :: Either AWS.Error a -> Amazon a
+throwA = either (throwM . GeneralError) return
+
+execCatch ::
+  (AWSRequest a, AWS.HasEnv r, MonadUnliftIO m, MonadCatch m, MonadThrow m) =>
+  r ->
+  a ->
+  m (Either AWS.Error (Rs a))
+execCatch e cmd =
+  runResourceT . AWST.runAWST e
+    $ AWST.trying AWS._Error
+    $ AWST.send cmd
+
+exec ::
+  (AWSRequest a, AWS.HasEnv r, MonadUnliftIO m, MonadCatch m, MonadThrow m) =>
+  r ->
+  a ->
+  m (Rs a)
+exec e cmd = execCatch e cmd >>= either (throwM . GeneralError) return
```

Comment:
These are not used and have functions with the same name on the S3 module; perhaps move those functions here since they are generic and can be used in other modules (although they are not)

Reply:
It was not quite as easy as copying them, because of the module dependencies (`Cargohold.AWS` doesn't know about `App` and `Env`), but here it is: https://github.com/wireapp/wire-server/pull/1157/commits/959adbcd2455ad895e392d1a656fa10b32dd9336